framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.TypeSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Type.ArrayType;
+import com.sun.tools.javac.code.Type.ClassType;
+
+/**
+ * This class writes inferred types for fields, method return types and method
+ * parameters to a .jaif file.  Calling an update* method
+ * ({@link SignatureInferenceScenes#updateFieldTypeInScene updateFieldTypeInScene},
+ * {@link SignatureInferenceScenes#updateMethodParameterTypeInScene updateMethodParameterTypeInScene} or
+ * {@link SignatureInferenceScenes#updateMethodReturnTypeInScene updateMethodReturnTypeInScene}) 
+ * reads the currently-stored type, if any, and replaces it by the LUB of
+ * it and the update method's argument.
+ *
+ * This class does not perform inference for elements that have explicit
+ * annotations - an explicitly annotated field type, method return type or
+ * method parameter type will not have its respective inferred type written
+ * into a .jaif file.
+ *
+ * @author pbsf
+ */
+public class SignatureInferenceScenes {
+
+    /**
+     * Path to where .jaif files will be written to and read from.
+     * This path is relative to where the CF's javac command is executed.
+     */
+    public final static String JAIF_FILES_PATH = "build/jaif-files/";
+
+    /** Maps file paths (Strings) to Scenes. */
+    private static Map<String, AScene> scenes = new HashMap<String, AScene>();
+
+    /** Set containing all Scenes that were modified in the current ClassTree.
+     * Modifying a Scene means adding (or changing) a type annotation for a
+     * field, method return type or method parameter type in the Scene.
+     * (Scenes are modified by the method
+     * {@link SignatureInferenceScenes#updateAnnotationSetInScene updateAnnotationSetInScene)}
+     */
+    private static Set<String> modifiedScenes = new HashSet<String>();
+
+    /**
+     * Returns the Scene related to a .jaif file path passed as input.
+     */
+    private static AScene getScene(String jaifPath) {
+        if (!scenes.containsKey(jaifPath)) {
+            File jaifFile = new File(jaifPath);
+            AScene scene = new AScene();
+            if (jaifFile.exists()) {
+                try {
+                    IndexFileParser.parseFile(jaifPath, scene);
+                } catch (IOException e) {
+                    ErrorReporter.errorAbort("Could not open file in: " + jaifPath + "."
+                            + " Exception message: " + e.getMessage());
+                }
+            }
+            scenes.put(jaifPath, scene);
+        }
+        return scenes.get(jaifPath);
+    }
+
+    /**
+     * Clears the set of modified scenes.
+     * (Scenes are modified by the method
+     * {@link SignatureInferenceScenes#updateAnnotationSetInScene updateAnnotationSetInScene)}
+     */
+    public static void clearModifiedScenes() {
```

Comment:
Why is this method public?  Should it be called by writeScenesToJaif instead?  It seems less error-prone to have the class maintain its modified state rather than forcing the client to do so.